test/framework/src/main/java/org/elasticsearch/index/engine/EngineTestCase.java

```diff
@@ -1259,4 +1276,76 @@ public static long getInFlightDocCount(Engine engine) {
     public static void assertNoInFlightDocuments(Engine engine) throws Exception {
         assertBusy(() -> assertThat(getInFlightDocCount(engine), equalTo(0L)));
     }
+
+    public static final class MatchingDirectoryReader extends FilterDirectoryReader {
+        private final Query query;
+
+        public MatchingDirectoryReader(DirectoryReader in, Query query) throws IOException {
+            super(in, new SubReaderWrapper() {
+                @Override
+                public LeafReader wrap(LeafReader leaf) {
+                    try {
+                        final IndexSearcher searcher = new IndexSearcher(leaf);
+                        final Weight weight = searcher.createWeight(query, ScoreMode.COMPLETE_NO_SCORES, 1.0f);
+                        final Scorer scorer = weight.scorer(leaf.getContext());
+                        final DocIdSetIterator iterator = scorer != null ? scorer.iterator() : null;
+                        final FixedBitSet liveDocs = new FixedBitSet(leaf.maxDoc());
+                        if (iterator != null) {
+                            for (int docId = iterator.nextDoc(); docId != DocIdSetIterator.NO_MORE_DOCS; docId = iterator.nextDoc()) {
+                                if (leaf.getLiveDocs() == null || leaf.getLiveDocs().get(docId)) {
+                                    liveDocs.set(docId);
+                                }
+                            }
+                        }
+                        return new FilterLeafReader(leaf) {
+                            @Override
+                            public Bits getLiveDocs() {
+                                return liveDocs;
+                            }
+
+                            @Override
+                            public CacheHelper getCoreCacheHelper() {
+                                return leaf.getCoreCacheHelper();
+                            }
+
+                            @Override
+                            public CacheHelper getReaderCacheHelper() {
+                                return null; // modify liveDocs
+                            }
+                        };
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                }
+            });
+            this.query = query;
+        }
+
+        @Override
+        protected DirectoryReader doWrapDirectoryReader(DirectoryReader in) throws IOException {
+            return new MatchingDirectoryReader(in, query);
+        }
+
+        @Override
+        public CacheHelper getReaderCacheHelper() {
+            return in.getReaderCacheHelper();
```

Comment:
Might mostly be for my understanding, but since the wrapped directory reader could return different results from the inner reader, should we not return null here?

Reply:
Good catch. We should return null here as we modify the liveDocs. I addressed it in https://github.com/elastic/elasticsearch/pull/64504/commits/eb968e4400f0dda44fb88f65952aaa31cdaad3d8.

Reply:
I had to reverted this and added a comment instead. Please see: https://github.com/elastic/elasticsearch/pull/64504/commits/e176bf2a86bfebac45b18a7e4dff1b53dd09f426